framework.go

```diff
@@ -0,0 +1,171 @@
+func New() (*Framework, error) {
+	mcKubeconfig, ok := os.LookupEnv(KubeconfigEnvVar)
+	if !ok {
+		return nil, fmt.Errorf("no %s set", KubeconfigEnvVar)
+	}
+
+	framework := NewWithKubeconfig(mcKubeconfig)
+
+	// for _, envVar := range os.Environ() {
+	// 	if strings.HasPrefix(envVar, "E2E_WC_") {
+	// 		// TODO: Initialize workload cluster
+	// 	}
+	// }
+
+	return framework, nil
+}
+
+// NewWithKubeconfig generates a new framework initialised with the Management Cluster provided as a Kubeconfig
+func NewWithKubeconfig(kubeconfigPath string) *Framework {
+	mcClient, err := client.New(kubeconfigPath)
+	if err != nil {
+		panic(err)
+	}
+
+	return &Framework{
+		kubeconfigPath: kubeconfigPath,
+		mcClient:       mcClient,
+		wcClients:      map[string]*client.Client{},
+	}
+}
+
+// MC returns an initialized client for the Management Cluster
+func (f *Framework) MC() *client.Client {
+	return f.mcClient
+}
+
+// WC returns an initialized client for the Workload Cluster matching the given name.
+// If no Workload Cluster is found matching the given name an error is returned.
+func (f *Framework) WC(clusterName string) (*client.Client, error) {
+	c, ok := f.wcClients[clusterName]
+	if !ok {
+		return nil, fmt.Errorf("workload cluster not found for name %s", clusterName)
+	}
+	return c, nil
+}
+
+// ApplyCluster takes a Cluster object, applies it to the MC in the correct order and then waits for a valid Kubeconfig to be available
+//
+// A timeout can be provided via the given `ctx` value by using `context.WithTimeout()`
+func (f *Framework) ApplyCluster(ctx context.Context, cluster *application.Cluster) (*client.Client, error) {
+	clusterApplication, clusterCM, defaultAppsApplication, defaultAppsCM, err := cluster.Build()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build cluster app: %v", err)
+	}
+
+	// Apply Cluster resources
+	if err := f.MC().Client.Create(ctx, clusterCM); err != nil {
+		return nil, fmt.Errorf("failed to apply cluster configmap: %v", err)
+	}
+	if err := f.MC().Client.Create(ctx, clusterApplication); err != nil {
+		return nil, fmt.Errorf("failed to apply cluster app CR: %v", err)
+	}
+	// Apply Default Apps resources
+	if err := f.MC().Client.Create(ctx, defaultAppsCM); err != nil {
+		return nil, fmt.Errorf("failed to apply default-apps configmap: %v", err)
+	}
+	if err := f.MC().Client.Create(ctx, defaultAppsApplication); err != nil {
+		return nil, fmt.Errorf("failed to apply default-apps app CR: %v", err)
+	}
+
+	return f.WaitForClusterReady(ctx, cluster.Name, cluster.Namespace)
+}
+
+// WaitForClusterReady watches for a Kubeconfig secret to be created on the MC and then waits until that cluster's api-server response successfully
+//
+// A timeout can be provided via the given `ctx` value by using `context.WithTimeout()`
+func (f *Framework) WaitForClusterReady(ctx context.Context, clusterName string, namespace string) (*client.Client, error) {
```

Comment:
if the framework is built with `ginkgo` and `gomega` in mind, why don't we rely on [`Eventually` and `Consistently`](https://onsi.github.io/gomega/#making-asynchronous-assertions)?

Reply:
Currently, the framework is designed to work with ginkgo but doesn't strictly require it. The only current exception to this is the logger function which isn't perfect right now anyway. 

Do you think it's best to only target `ginkgo` testing for this framework?

Reply:
Can this be considered future improvements? I'm currently not sure how I'd rework the code to only return a boolean and not an error when needed also.

Reply:
Ginkgo actually expects the error and will fail if it isn't nil. [Docs](https://pkg.go.dev/github.com/onsi/gomega#Eventually) :
> If multiple values are returned by the function, Eventually will pass the first value to the matcher and require that all others are zero-valued. This allows you to pass Eventually a function that returns a value and an error - a common pattern in Go. 

The only caveat is that it needs to return a function, since the only was to pass the parameters is to do it with a closure. So something like:
```go
func (f *Framework) IsClusterReady(ctx context.Context, clusterName string, namespace string) func(g gomega.Gomega) (*client.Client, error) {
   return func(g gomega.Gomega) (*client.Client, error) { // the gomega.Gomega part isn't necessary it can also be with empty params. It's there to correct for line numbers when the test fails. Otherwise it just tells you that it failed in the eventually and not the exact line number of the anonymous function  
   .... // use ctx, clusterName and namespace here
   }
}
```
But yeah it can also be a future improvement.

Reply:
I believe targeting `ginkgo` would make things easier/simpler, but I'm fine with going the agnostic way if you think that's better :+1: